src/plans/constrained_plan.jl

```diff
@@ -343,6 +362,182 @@ end
+
+    LagrangianCost(co, μ, λ)
+
+Create a functor for the Lagrangian with fixed dual variables.
+
+# Example
+
+When you directly want to evaluate the Lagrangian ``\mathcal L``
+you can also call
+
+```
+LagrangianCost(co, μ, λ)(M,p)
+```
+"""
+mutable struct LagrangianCost{CO,T}
+    co::CO
+    μ::T
+    λ::T
+end
+function (lc::LagrangianCost)(M, p)
+    c = get_cost(M, lc.co, p)
+    g = get_inequality_constraint(M, lc.co, p, :)
+    h = get_equality_constraint(M, lc.co, p, :)
+    (length(g) > 0) && (c += sum(lc.μ .* g))
+    (length(h) > 0) && (c += sum(lc.λ .* h))
+    return c
+end
+function set_manopt_parameter!(lc::LagrangianCost, ::Val{:μ}, μ)
+    lc.μ = μ
+    return lc
+end
+function set_manopt_parameter!(lc::LagrangianCost, ::Val{:λ}, λ)
+    lc.λ = λ
+    return lc
+end
+function show(io::IO, lc::LagrangianCost)
+    return print(io, "LagrangianCost\n\twith μ=$(lc.μ), λ=$(lc.λ)")
+end
+
+@doc raw"""
+    LagrangianGradient{CO, V, T}
+
+Implement the gradient of the Lagrangian of a [`ConstrainedManifoldObjective`](@ref) `co`.
+
+```math
+\operatorname{grad} \mathcal L(p, μ, λ)
+= \operatorname{grad} f(p) +  \sum_{i=1}^m μ_i \operatorname{grad} g_i(p) + \sum_{j=1}^n λ_j \operatorname{grad} h_j(p)
+```
+
+# Fields
+
+* `co::CO`, `μ::T`, `λ::T` as mentioned, where `T` represents a vector type.
+
+# Constructor
+
+    LagrangianGradient(co, μ, λ)
+
+Create a functor for the Lagrangian with fixed dual variables.
+
+# Example
+
+When you directly want to evaluate the gradient of the Lagrangian ``\operatorname{grad} \mathcal L``
+you can also call `LagrangianGradient(co, μ, λ)(M,p)` or `LagrangianGradient(co, μ, λ)(M,X,p)` for the in-place variant.
+"""
+mutable struct LagrangianGradient{CO,T}
+    co::CO
+    μ::T
+    λ::T
+end
+function (lg::LagrangianGradient)(M, p)
+    X = zero_vector(M, p)
+    return lg(M, X, p)
+end
+function (lg::LagrangianGradient)(M, X, p)
+    Y = copy(M, p, X)
+    get_gradient!(M, X, lg.co, p)
+    m = inequality_constraints_length(lg.co)
+    n = equality_constraints_length(lg.co)
+    for i in 1:m
+        get_grad_inequality_constraint!(M, Y, lg.co, p, i)
+        copyto!(M, X, p, X + lg.μ[i] * Y)
+    end
+    for j in 1:n
+        get_grad_equality_constraint!(M, Y, lg.co, p, j)
+        copyto!(M, X, p, X + lg.λ[j] * Y)
+    end
+    return X
+end
+function set_manopt_parameter!(lc::LagrangianGradient, ::Val{:μ}, μ)
+    lc.μ = μ
+    return lc
+end
+function set_manopt_parameter!(lc::LagrangianGradient, ::Val{:λ}, λ)
+    lc.λ = λ
+    return lc
+end
+function show(io::IO, lg::LagrangianGradient)
+    return print(io, "LagrangianGradient\n\twith μ=$(lg.μ), λ=$(lg.λ)")
+end
+
+@doc raw"""
+    LagrangianHessian{CO, V, T}
+
+Implement the Hessian of the Lagrangian of a [`ConstrainedManifoldObjective`](@ref) `co`.
+
+```math
+\operatorname{Hess} \mathcal L(p, μ, λ)[X]
+= \operatorname{Hess} f(p) +  \sum_{i=1}^m μ_i \operatorname{Hess} g_i(p)[X] + \sum_{j=1}^n λ_j \operatorname{Hess} h_j(p)[X]
```

Comment:
A small note: for Hessians and gradients of multi-argument functions like `L` I like to have an indication with respect to which variable it is taken (like in the original paper for IPM). So for example `\operatorname{Hess}_p` instead of just `\operatorname{Hess}`. I think it makes it easier to read.

Reply:
Similarly the text could be more clear that it's the Hessian with respect to `p`.

Reply:
Sure, that sounds good. I will carefully check that when working on this.